userspace/lib/usbip_forward.c

```diff
@@ -368,44 +369,53 @@ read_completion(DWORD errcode, DWORD nread, LPOVERLAPPED lpOverlapped)
 	devbuf_t	*rbuff;
 
 	rbuff = (devbuf_t *)lpOverlapped->hEvent;
-
-	if (errcode == 0) {
-		rbuff->offp += nread;
-		if (nread == 0)
-			rbuff->invalid = TRUE;
+	if (rbuff != NULL)
```

Comment:
The case that rbuff is NULL would never happen. I'll drop this line checking rbuff when this commit is cherry-picked into the master branch.

Reply:
Hopefully, this would never happen. However, while testing, when usbip failed on attach, it would bring usbipd down with it. I know this is specific to the fact that I'm modifying usbip to have it more like Linux' version, but we should never assume that this will never happen in a real life scenario (a connection may drop somewhere on the network). Bringing usbipd down so easily should not happen. Also, it doesn't cost much to have this check in place. Please, reconsider keeping it.

Reply:
I have confirmed that usbipd incurs seg fault at the rbuff check line when a forwarding routine of usbipd stops as you pointed out. But this fault is raised by not only disconnection and also detaching. In my opinion, it is due to that the forwarding routine does not clean up asynchronous I/O even though CancelIoEx is invoked before the forwarding routine exits. This problem should be another github issue.
Anyway, checking rbuff does not seem to be needed because read_completion should not be called after disconnection or detaching.

Reply:
Yes, I had not tested with other actions, but you are confirming what I thought. I'll open a different issue then so we can track this bug.

Reply:
I attempted to evaluate whether rbuff is null or not but found that it's never null. I think the issue is rbuff has been freed in the thread teardown however the callback still has a reference to it. This makes the pointer point to memory we no longer own, causing the crash.